tests/subsys/bluetooth/fast_pair/storage/factory_reset/src/main.c

```diff
@@ -0,0 +1,424 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/ztest.h>
+#include <zephyr/settings/settings.h>
+#include <zephyr/sys/reboot.h>
+
+#include <bluetooth/services/fast_pair.h>
+
+#include "fp_storage_ak.h"
+#include "fp_storage_ak_priv.h"
+#include "fp_storage_pn.h"
+#include "fp_storage_pn_priv.h"
+#include "fp_storage_manager.h"
+#include "fp_storage_manager_priv.h"
+
+#include "common_utils.h"
+
+
+#define ACCOUNT_KEY_MAX_CNT CONFIG_BT_FAST_PAIR_STORAGE_ACCOUNT_KEY_MAX
+
+#define SETTINGS_RESET_TEST_SUBTREE_NAME "fp_reset_test"
+#define SETTINGS_STAGE_KEY_NAME "stage"
+#define SETTINGS_NAME_SEPARATOR_STR "/"
+#define SETTINGS_RESET_TEST_STAGE_FULL_NAME \
+	(SETTINGS_RESET_TEST_SUBTREE_NAME SETTINGS_NAME_SEPARATOR_STR SETTINGS_STAGE_KEY_NAME)
+
+BUILD_ASSERT(SETTINGS_NAME_SEPARATOR == '/');
+
+enum reset_test_stage {
+	RESET_TEST_STAGE_NORMAL,
+	RESET_TEST_STAGE_INTERRUPT,
+	RESET_TEST_STAGE_INTERRUPT_FLOW_2,
+};
+
+static enum reset_test_stage test_stage = RESET_TEST_STAGE_NORMAL;
+
+static int settings_set_err;
+
+static bool storage_reset_reboot;
+
+
+static int settings_load_stage(size_t len, settings_read_cb read_cb, void *cb_arg)
+{
+	int rc;
+
+	if (len > sizeof(test_stage)) {
+		return -EINVAL;
+	}
+
+	rc = read_cb(cb_arg, &test_stage, len);
+	if (rc < 0) {
+		return rc;
+	}
+
+	if (rc != len) {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int fp_reset_test_settings_set(const char *name, size_t len, settings_read_cb read_cb,
+				      void *cb_arg)
+{
+	int err = 0;
+	static const char *key = SETTINGS_STAGE_KEY_NAME;
+
+	if (!strncmp(name, key, sizeof(SETTINGS_STAGE_KEY_NAME))) {
+		err = settings_load_stage(len, read_cb, cb_arg);
+	} else {
+		err = -ENOENT;
+	}
+
+	/* The first reported settings set error will be remembered by the module.
+	 * The error will then be propagated by the commit callback.
+	 * Errors returned in the settings set callback are not propagated further.
+	 */
+	if (err && !settings_set_err) {
+		settings_set_err = err;
+	}
+
+	return err;
+}
+
+static int fp_reset_test_settings_commit(void)
+{
+	if (settings_set_err) {
+		return settings_set_err;
+	}
+
+	return 0;
+}
+
+SETTINGS_STATIC_HANDLER_DEFINE(fp_reset_test, SETTINGS_RESET_TEST_SUBTREE_NAME, NULL,
+			       fp_reset_test_settings_set, fp_reset_test_settings_commit, NULL);
+
+static int test_fp_storage_reset_perform(void)
+{
+	if (storage_reset_reboot) {
+		sys_reboot(SYS_REBOOT_WARM);
```

Comment:
Do we really need to reboot the device or is it enough to simulate this?

To simulate the reboot, it is enough to clean the RAM memory and call ```settings_load```. Am I missing something?

I think it would simplify the test logic a lot.

Reply:
To simulate reboot we also need to somehow interrupt the execution of the `bt_fast_pair_factory_reset` function in the middle of it and I think the cleanest and safest solution is to actually perform reboot. We've discussed it with @MarekPieta and that looked like the best way to do it, but maybe you have some other idea?

Reply:
Discussed offline. Seems like there is no better solution than the mentioned reboot.